### PR TITLE
Patch manager

### PR DESCRIPTION
**Pull request by rocketz**
PatchManager is an easy way to patch in return instructions (with nop delay slot) and also just a single NOP.

There is a window where you can see your patches and chose to activate/deactive and delete them.

Clicking on the address in this window will take you there in the Assembly window.

### CODE REVIEW

File: src/core/patchmanager.cc [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/***************************************************************************
+ *   Copyright (C) 2022 PCSX-Redux authors                                 *
```
Comment (johnbaumann) [inline src/core/patchmanager.cc:2]: Need to bump the copyright year

File: src/core/patchmanager.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/***************************************************************************
+ *   Copyright (C) 2022 PCSX-Redux authors                                 *
```
Comment (johnbaumann) [inline src/core/patchmanager.h:2]: Here too

File: src/core/patchmanager.cc (hunk 1 above)
Comment (johnbaumann) [inline src/core/patchmanager.cc:2]: Need to update the copyright year

Comment (rocketz) [reply src/core/patchmanager.cc]: Fixed + 2 others

File: src/gui/widgets/patches.cc [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            ImGui::TableSetColumnIndex(3);
+            ImGui::TextUnformatted(patch.type == PCSX::PatchManager::Patch::Type::Return ? "Return" : "NOP");
+        }
+        ImGui::EndTable();
+
+        if (deleteIndex != -1) {
+            patchManager.deletePatch(deleteIndex);
+        }
+
+        if (ImGui::Button("Activate All")) {
```
Comment (nicolasnoble) [inline src/gui/widgets/patches.cc:102]:
```suggestion
        if (ImGui::Button(_("Activate All"))) {
```

File: src/gui/widgets/patches.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            patchManager.activateAll();
+        }
+
+        ImGui::SameLine();
+        if (ImGui::Button("Deactivate All")) {
```
Comment (nicolasnoble) [inline src/gui/widgets/patches.cc:107]:
```suggestion
        if (ImGui::Button(_("Deactivate All"))) {
```

File: src/gui/widgets/patches.cc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            patchManager.deactivateAll();
+        }
+
+        ImGui::SameLine();
+        if (ImGui::Button("Delete All")) {
```
Comment (nicolasnoble) [inline src/gui/widgets/patches.cc:112]:
```suggestion
        if (ImGui::Button(_("Delete All"))) {
```

**Review by nicolasnoble — COMMENTED** [review]
Don't freak out from the volume of comments I've left, it's mostly cosmetic. There's only a couple of functional nits which may need some rework, namely the checkbox label, and the cache invalidation call.

Great work, thanks for that!

File: src/gui/widgets/patches.cc [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+void PCSX::Widgets::Patches::draw(const char* title) {
+    ImGui::SetNextWindowPos(ImVec2(520, 30), ImGuiCond_FirstUseEver);
+    ImGui::SetNextWindowSize(ImVec2(600, 500), ImGuiCond_FirstUseEver);
+    if (!ImGui::Begin(title, &m_show)) {
+        ImGui::End();
+        return;
+    }
+
+    if (PCSX::g_emulator->m_cpu->isDynarec()) {
+        ImGui::TextUnformatted(_("Patching is only available in Interpreted CPU mode"));
```
[the remark this answers is not in the document]
Comment (rocketz) [reply src/gui/widgets/patches.cc:34]: Ok, I'm calling this now. Not sure how useful patches when not reverse-engineering and using the debugger though.

File: src/gui/widgets/patches.cc [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                if (ImGui::Button("Delete")) {
+                    deleteIndex = row;
+                    ImGui::CloseCurrentPopup();
+                }
+                ImGui::EndPopup();
+            }
+
+            ImGui::TableSetColumnIndex(2);
+            ImGui::PushID(row);
+            if (ImGui::Checkbox("", &patch.active))
```
[the remark this answers is not in the document]
Comment (rocketz) [reply src/gui/widgets/patches.cc:77]: Yeah, that should work fine afaik

**Review by nicolasnoble — APPROVED** [review]

**Pull request merged**